backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/ArtifactHandler.java

```diff
@@ -0,0 +1,223 @@
+                }
+
+                if (errorMessage != null) {
+                  throw new InvalidArgumentException(errorMessage);
+                }
+              }
+            },
+            executor);
+
+    // Log
+    for (final var artifact : artifacts) {
+      currentFuture =
+          currentFuture.thenCompose(
+              unused ->
+                  // Insert into ARTIFACT table
+                  jdbi.useHandle(
+                      handle -> {
+                        handle
+                            .createQuery(
+                                "select id from "
+                                    + getTableName()
+                                    + " where entity_name=:entity_name and field_type=:field_type and ar_key=:key and "
+                                    + entityIdReferenceColumn
+                                    + "=:entity_id")
+                            .bind("key", artifact.getKey())
+                            .bind("field_type", fieldType)
+                            .bind("entity_name", entityName)
+                            .bind("entity_id", entityId)
+                            .mapTo(Long.class)
+                            .findOne()
+                            .ifPresent(
+                                present -> {
+                                  throw new AlreadyExistsException(
+                                      "Key '" + artifact.getKey() + "' already exists");
+                                });
+
+                        handle
+                            .createUpdate(
+                                "insert into "
+                                    + getTableName()
+                                    + " (entity_name, field_type, ar_key, ar_path, artifact_type, path_only, linked_artifact_id, filename_extension, store_type_path,"
+                                    + entityIdReferenceColumn
+                                    + ") "
+                                    + "values (:entity_name, :field_type, :key, :path, :type,:path_only,:linked_artifact_id,:filename_extension,:store_type_path, :entity_id)")
+                            .bind("key", artifact.getKey())
+                            .bind("path", artifact.getPath())
+                            .bind("type", artifact.getArtifactTypeValue())
+                            .bind("path_only", artifact.getPathOnly())
+                            .bind("linked_artifact_id", artifact.getLinkedArtifactId())
+                            .bind("filename_extension", artifact.getFilenameExtension())
+                            .bind(
+                                "store_type_path",
+                                !artifact.getPathOnly()
+                                    ? Config.getInstance().artifactStoreConfig.storeTypePathPrefix()
+                                        + artifact.getPath()
+                                    : "")
```

Comment:
Extract to avoid complex inline logic in the middle of a clean `bind` sequence.

Reply:
Extracted to a var